File: runtime/android/NexaDevActions.kt
package __NEXA_PACKAGE__

import org.json.JSONArray
import org.json.JSONObject
import kotlinx.coroutines.CancellationException

internal enum class NexaDevActionFlow { Normal, Break, Continue }

internal fun Any?.orNullValue(): Any = this ?: JSONObject.NULL

internal fun NexaDevStateStore.perform(actions: JSONArray, scope: String, locals: Map<String, Any>) {
    performActions(actions, scope, locals)
}

internal fun NexaDevStateStore.performActions(actions: JSONArray, scope: String, locals: Map<String, Any>): NexaDevActionFlow {
    for (index in 0 until actions.length()) {
        val raw = actions.opt(index)
        if (raw is String) {
            if (raw == "Break") return NexaDevActionFlow.Break
            if (raw == "Continue") return NexaDevActionFlow.Continue
            continue
        }
        val action = raw as? JSONObject ?: continue
        val assign = action.optJSONObject("Assign")
        if (assign != null) {
            val name = assign.optString("name")
            val nextValue = evaluate(assign.opt("value"), locals, scope)
            setState(name, nextValue, scope)
            moduleRevision++
            continue
        }
        val expression = action.opt("Expression")
        if (expression != null) {
            evaluate(expression, locals, scope)
            continue
        }
        val branch = action.optJSONObject("If")
        if (branch != null) {
            val selected = if (truthy(evaluate(branch.opt("condition"), locals, scope))) {
                branch.optJSONArray("then_branch")
            } else {
                branch.optJSONArray("else_branch")
            }
            val flow = performActions(selected ?: JSONArray(), scope, locals)
            if (flow != NexaDevActionFlow.Normal) return flow
            continue
        }
        val loop = action.optJSONObject("For")
        if (loop != null) {
            val name = loop.optString("name")
            val iterable = evaluate(loop.opt("iterable"), locals, scope)
            val items = when (iterable) {
                is List<*> -> iterable
                is Set<*> -> iterable.toList()
                else -> emptyList<Any?>()
            }
            iteration@ for (item in items) {
                val nextLocals = locals.toMutableMap()
                nextLocals[name] = item.orNullValue()
                when (performActions(loop.optJSONArray("body") ?: JSONArray(), scope, nextLocals)) {
                    NexaDevActionFlow.Break -> break@iteration
                    NexaDevActionFlow.Continue, NexaDevActionFlow.Normal -> Unit
                }
            }
            continue
        }
        val mapLoop = action.optJSONObject("ForMap")
        if (mapLoop != null) {
            val keyName = mapLoop.optString("key_name")
            val valueName = mapLoop.optString("value_name")
            val iterable = evaluate(mapLoop.opt("iterable"), locals, scope)
            val map = iterable as? Map<*, *> ?: emptyMap<Any?, Any?>()
            iteration@ for ((key, value) in map) {
                val nextLocals = locals.toMutableMap()
                nextLocals[keyName] = key.orNullValue()
                nextLocals[valueName] = value.orNullValue()
                when (performActions(mapLoop.optJSONArray("body") ?: JSONArray(), scope, nextLocals)) {
                    NexaDevActionFlow.Break -> break@iteration
                    NexaDevActionFlow.Continue, NexaDevActionFlow.Normal -> Unit
                }
            }
            continue
        }
        val whileLoop = action.optJSONObject("While")
        if (whileLoop != null) {
            var iterations = 0
            iteration@ while (iterations++ < 10_000 && truthy(evaluate(whileLoop.opt("condition"), locals, scope))) {
                when (performActions(whileLoop.optJSONArray("body") ?: JSONArray(), scope, locals)) {
                    NexaDevActionFlow.Break -> break@iteration
                    NexaDevActionFlow.Continue, NexaDevActionFlow.Normal -> Unit
                }
            }
            continue
        }
        val mutation = action.optJSONObject("CollectionMutation")
        if (mutation != null) {
            mutateCollection(mutation, scope, locals)
            continue
        }
        val tryCatch = action.optJSONObject("TryCatch")
        if (tryCatch != null) {
            val flow = performActions(tryCatch.optJSONArray("body") ?: JSONArray(), scope, locals)
            if (flow != NexaDevActionFlow.Normal) return flow
            continue
        }
        if (action.has("Break")) return NexaDevActionFlow.Break
        if (action.has("Continue")) return NexaDevActionFlow.Continue
    }
    return NexaDevActionFlow.Normal
}

internal fun NexaDevStateStore.mutateCollection(mutation: JSONObject, scope: String, locals: Map<String, Any>) {
    val name = mutation.optString("name")
    val rawArguments = mutation.optJSONArray("arguments") ?: JSONArray()
    val arguments = (0 until rawArguments.length()).map { evaluate(rawArguments.opt(it), locals, scope) }
    when (mutation.optString("operation")) {
        "ArrayAppend" -> {
            val target = (state(name, scope) as? List<*>)?.toMutableList() ?: mutableListOf<Any?>()
            if (arguments.isNotEmpty()) target.add(arguments[0])
            setState(name, target, scope)
            moduleRevision++
        }
        "ArrayRemoveAt" -> {
            val target = (state(name, scope) as? List<*>)?.toMutableList() ?: mutableListOf<Any?>()
            val index = (arguments.firstOrNull() as? Number)?.toInt() ?: -1
            if (index in 0 until target.size) target.removeAt(index)
            setState(name, target, scope)
            moduleRevision++
        }
        "SetInsert", "SetRemove" -> {
            val target = (state(name, scope) as? Set<*>)?.map { stringify(it ?: JSONObject.NULL) }?.toMutableSet() ?: mutableSetOf()
            val item = arguments.firstOrNull()?.let { stringify(it) }
            if (item != null) {
                if (mutation.optString("operation") == "SetInsert") target.add(item) else target.remove(item)
            }
            setState(name, target, scope)
            moduleRevision++
        }
        "MapSet", "MapRemove" -> {
            val target = (state(name, scope) as? Map<*, *>)?.mapKeys { stringify(it.key ?: JSONObject.NULL) }?.toMutableMap() ?: mutableMapOf()
            val key = arguments.firstOrNull()?.let { stringify(it) }
            if (key != null) {
                if (mutation.optString("operation") == "MapSet" && arguments.size > 1) {
                    target[key] = arguments[1]
                } else {
                    target.remove(key)
                }
            }
            setState(name, target, scope)
            moduleRevision++
        }
    }
}

internal suspend fun NexaDevStateStore.performAsync(actions: JSONArray, scope: String, locals: Map<String, Any>) {
    for (index in 0 until actions.length()) {
        val action = actions.optJSONObject(index) ?: continue
        val tryCatch = action.optJSONObject("TryCatch")
        if (tryCatch != null) {
            try {
                performAsync(tryCatch.optJSONArray("body") ?: JSONArray(), scope, locals)
            } catch (cancellation: CancellationException) {
                throw cancellation
            } catch (error: Throwable) {
                android.util.Log.e("NexaDevRuntime", "Async dev action failed", error)
                val catchBody = tryCatch.optJSONArray("catch_body")
                if (catchBody != null) {
                    performAsync(catchBody, scope, locals)
                } else {
                    throw error
                }
            }
            continue
        }
        val assign = action.optJSONObject("Assign")
        if (assign != null) {
            val name = assign.optString("name")
            val nextValue = evaluateAsync(assign.opt("value"), locals, scope)
            setState(name, nextValue, scope)
            moduleRevision++
            continue
        }
        val branch = action.optJSONObject("If")
        if (branch != null) {
            val conditionValue = evaluateAsync(branch.opt("condition"), locals, scope)
            val selected = if (truthy(conditionValue)) {
                branch.optJSONArray("then_branch")
            } else {
                branch.optJSONArray("else_branch")
            }
            performAsync(selected ?: JSONArray(), scope, locals)
            continue
        }
        val expression = action.opt("Expression")
        if (expression != null) {
            evaluateAsync(expression, locals, scope)
        }
    }
}

File: runtime/ios/NexaDevActions.swift
import Foundation

/// Flow control signal for synchronous Nexa dev action execution loops.
enum NexaDevActionFlow: Equatable {
    case normal
    case `break`
    case `continue`
}

@MainActor
extension NexaDevStateStore {
    func perform(_ actions: [Any], scope: String, locals: [String: Any]) {
        _ = performActions(actions, scope: scope, locals: locals)
    }

    @discardableResult
    func performActions(_ actions: [Any], scope: String, locals: [String: Any]) -> NexaDevActionFlow {
        for action in actions {
            if let unitVariant = action as? String {
                if unitVariant == "Break" { return .break }
                if unitVariant == "Continue" { return .continue }
                continue
            }
            guard let tagged = action as? [String: Any] else { continue }
            if let assignment = tagged["Assign"] as? [String: Any],
               let name = assignment["name"] as? String,
               let expression = assignment["value"] {
                setValue(name, value: evaluate(expression, locals: locals, scope: scope), scope: scope)
                revision += 1
            } else if let expression = tagged["Expression"] {
                _ = evaluate(expression, locals: locals, scope: scope)
            } else if let branch = tagged["If"] as? [String: Any],
                      let condition = branch["condition"] {
                let selected = truthy(evaluate(condition, locals: locals, scope: scope))
                    ? branch["then_branch"] as? [Any]
                    : branch["else_branch"] as? [Any]
                let flow = performActions(selected ?? [], scope: scope, locals: locals)
                if flow != .normal { return flow }
            } else if let loop = tagged["For"] as? [String: Any],
                      let name = loop["name"] as? String,
                      let iterableExpression = loop["iterable"] {
                let values = evaluate(iterableExpression, locals: locals, scope: scope)
                let items: [Any]
                if let array = values as? [Any] { items = array }
                else if let range = values as? Range<Int> { items = Array(range) }
                else if let range = values as? ClosedRange<Int> { items = Array(range) }
                else { items = [] }
                iterationLoop: for item in items {
                    var iterationLocals = locals
                    iterationLocals[name] = item
                    switch performActions(loop["body"] as? [Any] ?? [], scope: scope, locals: iterationLocals) {
                    case .break: break iterationLoop
                    case .normal, .continue: continue
                    }
                }
            } else if let loop = tagged["ForMap"] as? [String: Any],
                      let keyName = loop["key_name"] as? String,
                      let valueName = loop["value_name"] as? String,
                      let iterableExpression = loop["iterable"],
                      let map = evaluate(iterableExpression, locals: locals, scope: scope) as? [String: Any] {
                iterationLoop: for key in map.keys.sorted() {
                    var iterationLocals = locals
                    iterationLocals[keyName] = key
                    iterationLocals[valueName] = map[key] ?? NSNull()
                    switch performActions(loop["body"] as? [Any] ?? [], scope: scope, locals: iterationLocals) {
                    case .break: break iterationLoop
                    case .normal, .continue: continue
                    }
                }
            } else if let loop = tagged["While"] as? [String: Any], let condition = loop["condition"] {
                iterationLoop: for _ in 0..<10_000 {
                    guard truthy(evaluate(condition, locals: locals, scope: scope)) else { break }
                    switch performActions(loop["body"] as? [Any] ?? [], scope: scope, locals: locals) {
                    case .break: break iterationLoop
                    case .normal, .continue: continue
                    }
                }
            } else if let mutation = tagged["CollectionMutation"] as? [String: Any],
                      let name = mutation["name"] as? String {
                performCollectionMutation(mutation, name: name, scope: scope, locals: locals)
            } else if let tryCatch = tagged["TryCatch"] as? [String: Any] {
                let flow = performActions(tryCatch["body"] as? [Any] ?? [], scope: scope, locals: locals)
                if flow != .normal { return flow }
            } else if tagged["Break"] != nil {
                return .break
            } else if tagged["Continue"] != nil {
                return .continue
            }
        }
        return .normal
    }

    func performCollectionMutation(
        _ mutation: [String: Any],
        name: String,
        scope: String,
        locals: [String: Any]
    ) {
        let arguments = (mutation["arguments"] as? [Any] ?? []).map { evaluate($0, locals: locals, scope: scope) }
        switch mutation["operation"] as? String {
        case "ArrayAppend":
            var array = value(name, scope: scope) as? [Any] ?? []
            if let item = arguments.first { array.append(item) }
            setValue(name, value: array, scope: scope)
        case "ArrayRemoveAt":
            var array = value(name, scope: scope) as? [Any] ?? []
            if let index = (arguments.first as? NSNumber)?.intValue, array.indices.contains(index) { array.remove(at: index) }
            setValue(name, value: array, scope: scope)
        case "SetInsert", "SetRemove":
            var set = value(name, scope: scope) as? Set<String> ?? []
            if let item = arguments.first.map(stringify) {
                if mutation["operation"] as? String == "SetInsert" { set.insert(item) } else { set.remove(item) }
            }
            setValue(name, value: set, scope: scope)
        case "MapSet", "MapRemove":
            var map = value(name, scope: scope) as? [String: Any] ?? [:]
            if let key = arguments.first.map(stringify) {
                if mutation["operation"] as? String == "MapSet", arguments.count > 1 { map[key] = arguments[1] }
                else { map.removeValue(forKey: key) }
            }
            setValue(name, value: map, scope: scope)
        default: break
        }
    }

    func performAsync(_ actions: [Any], scope: String, locals: [String: Any]) async throws {
        for action in actions {
            guard let tagged = action as? [String: Any] else { continue }
            if let tryCatch = tagged["TryCatch"] as? [String: Any] {
                do {
                    try await performAsync(
                        tryCatch["body"] as? [Any] ?? [],
                        scope: scope,
                        locals: locals
                    )
                } catch is CancellationError {
                    throw CancellationError()
                } catch {
                    NSLog("NexaDevRuntime async dev action failed: %@", String(describing: error))
                    if let catchBody = tryCatch["catch_body"] as? [Any] {
                        try await performAsync(catchBody, scope: scope, locals: locals)
                    } else {
                        throw error
                    }
                }
            } else if let assignment = tagged["Assign"] as? [String: Any],
               let name = assignment["name"] as? String,
               let expression = assignment["value"] {
                let value = try await evaluateAsync(expression, locals: locals, scope: scope)
                setValue(name, value: value, scope: scope)
                revision += 1
            } else if let branch = tagged["If"] as? [String: Any],
                      let condition = branch["condition"] {
                let value = try await evaluateAsync(condition, locals: locals, scope: scope)
                let selected = truthy(value)
                    ? branch["then_branch"] as? [Any]
                    : branch["else_branch"] as? [Any]
                try await performAsync(selected ?? [], scope: scope, locals: locals)
            } else if let expression = tagged["Expression"] {
                _ = try await evaluateAsync(expression, locals: locals, scope: scope)
            }
        }
    }
}
